File: src/jvmTest/kotlin/org/snakeyaml/engine/issues/issue23/EmptyStringOutputTest.java
/*
 * Copyright (c) 2018, SnakeYAML
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */
package org.snakeyaml.engine.issues.issue23;

import org.junit.jupiter.api.Test;
import it.krzeminski.snakeyaml.engine.kmp.api.Dump;
import it.krzeminski.snakeyaml.engine.kmp.api.DumpSettings;
import it.krzeminski.snakeyaml.engine.kmp.api.StreamDataWriter;
import it.krzeminski.snakeyaml.engine.kmp.common.ScalarStyle;
import it.krzeminski.snakeyaml.engine.kmp.emitter.Emitter;
import it.krzeminski.snakeyaml.engine.kmp.events.DocumentStartEvent;
import it.krzeminski.snakeyaml.engine.kmp.events.ImplicitTuple;
import it.krzeminski.snakeyaml.engine.kmp.events.ScalarEvent;
import it.krzeminski.snakeyaml.engine.kmp.events.StreamStartEvent;

import java.io.StringWriter;
import java.util.HashMap;

import static org.junit.jupiter.api.Assertions.assertEquals;

@org.junit.jupiter.api.Tag("fast")
public class EmptyStringOutputTest {

  @Test
  void outputEmptyString() {
    Dump dumper = new Dump(DumpSettings.builder().build());
    String output = dumper.dumpToString("");
    assertEquals("''\n", output, "The output must NOT contain ---");
  }

  @Test
  void outputEmptyStringWithExplicitStart() {
    Dump dumper = new Dump(DumpSettings.builder().setExplicitStart(true).build());
    String output = dumper.dumpToString("");
    assertEquals("--- ''\n", output, "The output must contain ---");
  }

  @Test
  void outputEmptyStringWithEmitter() {
    assertEquals("---", dump(""), "The output must contain ---");
  }

  @Test
  void outputStringWithEmitter() {
    assertEquals("v1234512345", dump("v1234512345"), "The output must NOT contain ---");
  }

  private String dump(String value) {
    DumpSettings settings = DumpSettings.builder().build();
    MyWriter writer = new MyWriter();
    Emitter emitter = new Emitter(settings, writer);
    emitter.emit(new StreamStartEvent());
    emitter.emit(new DocumentStartEvent(false, null, new HashMap<>()));
    emitter.emit(new ScalarEvent(null, null, new ImplicitTuple(true, false),
        value, ScalarStyle.PLAIN, null, null));
    return writer.toString();
  }
}


class MyWriter extends StringWriter implements StreamDataWriter {

}

File: src/commonTest/kotlin/it/krzeminski/snakeyaml/engine/kmp/issues/issue23/EmptyStringOutputTest.kt
package it.krzeminski.snakeyaml.engine.kmp.issues.issue23

import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.shouldBe
import it.krzeminski.snakeyaml.engine.kmp.api.Dump
import it.krzeminski.snakeyaml.engine.kmp.api.DumpSettings
import it.krzeminski.snakeyaml.engine.kmp.api.StreamDataWriter
import it.krzeminski.snakeyaml.engine.kmp.common.ScalarStyle
import it.krzeminski.snakeyaml.engine.kmp.emitter.Emitter
import it.krzeminski.snakeyaml.engine.kmp.events.DocumentStartEvent
import it.krzeminski.snakeyaml.engine.kmp.events.ImplicitTuple
import it.krzeminski.snakeyaml.engine.kmp.events.ScalarEvent
import it.krzeminski.snakeyaml.engine.kmp.events.StreamStartEvent

class EmptyStringOutputTest : FunSpec({
    test("output empty string") {
        val dumper = Dump(DumpSettings())
        val output = dumper.dumpToString("")
        output shouldBe "''\n"
    }

    test("output empty string with explicit start") {
        val dumper = Dump(DumpSettings(isExplicitStart = true))
        val output = dumper.dumpToString("")
        output shouldBe "--- ''\n"
    }

    test("output empty string with emitter") {
        dump("") shouldBe "---"
    }

    test("output string with emitter") {
        dump("v1234512345") shouldBe "v1234512345"
    }
})

private fun dump(value: String): String {
    val settings = DumpSettings()
    val writer = MyWriter()
    val emitter = Emitter(settings, writer)
    emitter.emit(StreamStartEvent())
    emitter.emit(DocumentStartEvent(explicit = false, specVersion = null, tags = emptyMap()))
    emitter.emit(ScalarEvent(
        anchor = null,
        tag = null,
        implicit = ImplicitTuple(plain = true, nonPlain = false),
        value = value,
        scalarStyle = ScalarStyle.PLAIN,
        startMark = null,
        endMark = null,
    ))
    return writer.toString()
}

private class MyWriter : StreamDataWriter {
    private val content = StringBuilder()

    override fun write(str: String) {
        content.append(str)
    }

    override fun write(str: String, off: Int, len: Int) {
        content.append(str.substring(off, off + len))
    }

    override fun flush() {
        // No-op
    }

    override fun toString(): String = content.toString()
}
